src/piecewise.jl

```diff
@@ -0,0 +1,51 @@
+struct Constant{T}
+    value::T
+end
+
+(constant::Constant)(::Any) = constant.value
+(constant::Constant)(::Interval) = interval(constant.value)
+
+struct Piecewise{T <: Tuple}
+    pieces::T 
+    continuous::Bool
+end
+
+Piecewise(pairs... ; continuous = false) = Piecewise(pairs, continuous)
+
+domain(piecewise::Piecewise) = reduce(hull, subdomains(piecewise))
+subdomains(piecewise::Piecewise) = first.(piecewise.pieces)
+
+intersecting(X::Interval, Y::Interval) = !isempty_interval(intersect_interval(X, Y))
+
+function (piecewise::Piecewise)(X::Interval)
+    intersections = intersecting.(X, subdomains(piecewise))
+    dom = domain(piecewise)
+    !intersecting(X, dom) && return emptyinterval()
+    if !issubset_interval(X, dom)
+        dec = trv
+    elseif count(intersections) > 1
+        if piecewise.continuous
+            dec = dac
+        else
+            dec = def
+        end
+    else
+        dec = com
+    end
+
+    used_pieces = filter(piece -> intersecting(X, piece[1]), piecewise.pieces)
+    outputs = map(used_pieces) do (region, f)
+        S = IntervalArithmetic.setdecoration(intersect_interval(X, region), decoration(X))
+        return f(S)
+    end
+
+    dec = min(dec, minimum(decoration.(outputs)))
+    return IntervalArithmetic.setdecoration(reduce(hull, outputs), dec)
+end
+
+function (piecewise::Piecewise)(x::Real)
+    for (region, f) in piecewise.pieces
+        in_interval(x, region) && return f(x)
```

Comment:
This doesn't seem completely right: you need to take the union of `f(x)` over all the pieces that `x` overlaps with, I think.

Reply:
e.g. the line
```
    return reduce(hull, f(intersect_interval(X, region)) for (region, f) ∈ piecewise.pieces)

```
in my original implementation in the issue you limked to.

Reply:
This part is the implementation for non-interval real only, so I think it is fine, as it would be a bit awkward to return an interval for non-interval input.

As @OlivierHnt mentioned, this implementation allows ill-defined functions, but I don't think the answer is the interval-ification of everything.